src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -80,9 +84,15 @@ cdef class _ServicerContext:
         if self._metadata_sent:
             raise RuntimeError('Send initial metadata failed: already sent')
         else:
-            _send_initial_metadata(self._rpc_state, self._loop)
+            await _send_initial_metadata(self._rpc_state, metadata, self._loop)
```

Comment:
An open question, how much we would need to care about race conditions here? Theoretically, if we won't lock the whole `send_initial_metadata` function the user might call `_send_initial_metadata` more than once.

But as I said, just an open question. For me wouldn't be a blocker right now. //cc @lidizheng @gnossen

Reply:
The race condition is indeed a problem. We could 1) enforce the synchronization ourselves; 2) or educate users in docstring. In other implementations of gRPC, usually it is enforced by users (Golang, C++). For Python, I will vote for adding locks to make debugging simpler for our users.

Reply:
+1 to the locking.